src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs

```diff
@@ -81,6 +80,12 @@ public ulong Nonce
         /// <seealso cref="ProcessMessage(IncomingMessage)"/>
         private readonly EventLoopMessageListener<IncomingMessage> listener;
 
+        /// <summary>List of connected clients mapped by their unique identifiers.</summary>
+        private readonly ConcurrentDictionary<int, NetworkPeerClient> clientsById;
```

Comment:
What the purpose of this dic? beyond being a index mapping

Reply:
so before i believe the node had no idea about who is connected to it
so we did have some pending processing of the clients and on shutdown we could not wait until that processing is over. simply because there was no reference to those existing clients 

with this we are aware of who is connected to us and on shutdown we wait until everyone is gracefully disconnected, so then there will be no new incoming messages after shutdown

so should help with correct shutdown

Reply:
Don't we have `ConnectedNetworkPeers` for that?

Reply:
Unfortunately no, the network peer is connect much earlier before it is added to ConnectedNetworkPeers. That part is also a hot candidate for being rewritten, not an easy one though.

What we need here is something that binds the server with the client connections for the whole lifetime of the clients from were first moment to the very last moment WITHOUT any gaps. ConnectedNetworkPeers does not satisfy that requirement.